sqlglot/generator.py

```diff
@@ -2591,6 +2591,15 @@ def safedpipe_sql(self, expression: exp.SafeDPipe) -> str:
     def div_sql(self, expression: exp.Div) -> str:
         return self.binary(expression, "/")
 
+    def typeddiv_sql(self, expression: exp.TypedDiv) -> str:
+        l, r = expression.left, expression.right
+
+        if l.is_type(*exp.DataType.INTEGER_TYPES) and r.is_type(*exp.DataType.INTEGER_TYPES):
```

Comment:
right, so this cast should only happen if TYPE_DIVISION = False, as opposed to overriding the constants in dialect.py